res/controllers/Numark-Mixtrack-Platinum-scripts.js

```diff
@@ -0,0 +1,1415 @@
+
+        // update duration, time elapsed, and the spinner
+        var duration = engine.getValue(group, 'duration');
+        var position = engine.getValue(group, 'playposition');
+        if (duration == 0) position = 0;
+        MixtrackPlatinum.positionCallback(position, group, 'playposition');
+
+        // update bpm
+        var bpm = engine.getValue(group, 'bpm');
+        if (bpm != 0) MixtrackPlatinum.screenBpm(i + 1, Math.round(bpm * 100));
+
+        // keylock indicator
+        led(group, 'keylock', i, 0x0D);
+
+        // turn off bpm arrows
+        midi.sendShortMsg(0x80 | i, 0x0A, 0x00); // down arrow off
+        midi.sendShortMsg(0x80 | i, 0x09, 0x00); // up arrow off
+
+        // slip indicator
+        led(group, 'slip_enabled', i, 0x0F);
+
+        // initialize wheel mode (and leds)
+        MixtrackPlatinum.wheel[i] = EnableWheel;
+        midi.sendShortMsg(0x90 | i, 0x07, EnableWheel ? 0x7F : 0x01);
+
+        // initialize elapsed/remaining mode
+        MixtrackPlatinum.show_elapsed[i] = ShowTimeElapsed;
+        midi.sendShortMsg(0x90 | i, 0x46, ShowTimeElapsed ? 0x00 : 0x7F);
+    }
+
+    // zero vu meters
+    midi.sendShortMsg(0xBF, 0x44, 0);
+    midi.sendShortMsg(0xBF, 0x45, 0);
+
+    // setup position tracking
+    engine.connectControl("[Channel1]", "playposition", MixtrackPlatinum.positionCallback);
+    engine.connectControl("[Channel2]", "playposition", MixtrackPlatinum.positionCallback);
+    engine.connectControl("[Channel3]", "playposition", MixtrackPlatinum.positionCallback);
+    engine.connectControl("[Channel4]", "playposition", MixtrackPlatinum.positionCallback);
+
+    // setup bpm tracking
+    engine.connectControl("[Channel1]", "bpm", MixtrackPlatinum.bpmCallback);
+    engine.connectControl("[Channel2]", "bpm", MixtrackPlatinum.bpmCallback);
+    engine.connectControl("[Channel3]", "bpm", MixtrackPlatinum.bpmCallback);
+    engine.connectControl("[Channel4]", "bpm", MixtrackPlatinum.bpmCallback);
+
+    // setup vumeter tracking
+    engine.connectControl("[Channel1]", "VuMeter", MixtrackPlatinum.vuCallback);
+    engine.connectControl("[Channel2]", "VuMeter", MixtrackPlatinum.vuCallback);
+    engine.connectControl("[Channel3]", "VuMeter", MixtrackPlatinum.vuCallback);
+    engine.connectControl("[Channel4]", "VuMeter", MixtrackPlatinum.vuCallback);
+    engine.connectControl("[Master]", "VuMeterL", MixtrackPlatinum.vuCallback);
+    engine.connectControl("[Master]", "VuMeterR", MixtrackPlatinum.vuCallback);
+};
+
+MixtrackPlatinum.shutdown = function() {
+    // note: not all of this appears to be strictly necessary, things work fine
+    // with out this, but other software has been observed sending these led
+    // reset messages during shutdown. The last sysex message may be necessary
+    // to re-enable demo mode.
+
+    // turn off a bunch of channel specific leds
+    for (var i = 0; i < 4; ++i) {
+        // pfl/cue button leds
+        midi.sendShortMsg(0x90 | i, 0x1B, 0x01);
```

Comment:
Initially when I wrote this, I don't think the component system existed (or I wasn't aware of it). This code was copied over from that implementation. Some of these don't have components associated with them. It would be somewhat of a hassle to go through and figure out what's what and I don't see any benefit to it.

Reply:
Okay, leave as it is then